### PR TITLE
[DPE-4485] Split Kyuubi OCI image from Charmed Spark OCI image

### PR DESCRIPTION
**Pull request by theoctober19th**

### CODE REVIEW

File: build/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-ARG BASE_IMAGE=base-charmed-spark:latest
+ARG BASE_IMAGE=staged-charmed-spark:3.4.2
```
Comment (deusebio) [inline build/Dockerfile]: **niptick** Uhm, maybe I would keep using latest here. The number `3.4.2` can easily be not updated and not reflect actual workload. I don't think it is worth making sure that the "staged" image imported in docker as the proper version. It is just a intermediary output of the build process.

File: metadata.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+flavours:
```
Comment (deusebio) [inline metadata.yaml]: **praise** I really like the approach and the idea to use a `metadata.yaml` file to provide the details about the different images. I would possibly also add a key `build_file` where we provide the name of the dockerfile to be used. If needed, we could also use this file to custom the build of OCI images (e.g. adding tags to the image), but not needed right now. 

However in general, maybe we could make the build a bit more parametric if possible with respect to the flavour, such that if we need to add a new image, this only means:

1. Adding a Dockerbuild or rockcraft.yaml
2. Adding the YAML 

Hopefully without touching the makefile

File: Makefile [hunk 3]
```diff
@@ -11,146 +11,271 @@
...
+ROCK_NAME := $(shell yq .name rockcraft.yaml)
+
+# eg, 3.4.2
+SPARK_VERSION := $(shell yq .version rockcraft.yaml)
+
+# eg, 1.9.0
+KYUUBI_VERSION=$(shell yq .flavours.kyuubi.version metadata.yaml)
+
+# eg, 4.0.11
+JUPYTER_VERSION=$(shell yq .flavours.jupyter.version metadata.yaml)
```
Comment (deusebio) [inline Makefile]: I'm wondering whether we could make it more parametric, like that it takes:

```
$(shell yq .flavours.<flavour>.version metadata.yaml)
```

File: Makefile [hunk 4]
```diff
@@ -11,146 +11,271 @@
 # EXTERNAL VARIABLES
 # ======================
 
-REPOSITORY :=
-PREFIX :=
-TARGET := docker
+# The repository where the image is going to be hosted.
+# eg, ghcr.io/canonical [To be passed when you 'make' the recipe]
```
Comment (deusebio) [inline Makefile:15]: **praise** Huge praise for providing these comments, that helps understand the Makefile!!!!

**Review by deusebio — COMMENTED** [review]
I personally **REALLY** like the structure and the idea of using makefile logic + a YAML to describe/configure the building of the different images. 

My main comment is whether we could make the makefile more parametric (right now there are already a number of places where we hardcode JUPYTER, KYUUBI, etc), such that adding a new image would just be about 1. adding a new Dockerfile or rockcraft file and 2. adding the metadata in the YAML, such that at that point `make build FLAVOUR=xxxx` works out of the box. 

In this, I believe it would make sense if in the meta data we can provide details about:
1. The build backend (either `docker` or `rockcraft`) 
2. The manifest file to be used (e.g. `Dockerfile.my_flavour`)

Something like (but the name of the keys/syntax doesn't have to be this one, but it is just to provide the idea):
```
flavours:
    base:
        build:
            tool: rockcraft
            manifest: rockcraft.yaml
        ...
    jupyter:
        build:
            tool: docker
            manifest: build/Dockerfile.jupyter
        ...
```

I understand this may fall out of the scope of this PR. Maybe some small bits can be done in this context (e.g. adding the file specification in the `metadata.yaml` as well as adding the flavour for the base) while others may go in a dedicated ticket (e.g. making the makefile fully parametric)

File: build/Dockerfile (hunk 1 above)
Comment (theoctober19th) [reply build/Dockerfile]: I agree with this one. Updating it..

**Comment by theoctober19th** [comment]
@deusebio 
I agree that there are places where we could parameterize the Makefile. Theoretically, we may even be able to add some logic to the Makefile in such a way that only updating the manifest and adding Dockerfile would be enough for adding new images without the need of touching the Makefile.

However, I see a tradeoff here. The tradeoff being the readablity of Makefile vs ease of creating new image flavors. In the process of minimizing the usage of "jupyter" and "kyuubi" in the Makefile, we would use wildcards, conditionals and variable substitutions that would make Makefile even more convoluted (at least for someone who is not an expert on it). The reason I refactored it was because the previous version of Makefile was very convoluted with a lot of black magic happening here and there and it was really a pain to understand what was going on. I would personally not want to go back to the same state again. Also, I feel that we publish newer flavors of images every 6 months or so while we need to read and interact with Makefile a lot more frequent than that. Thus, I'd not mind it being a bit redundant if it becomes easier to read and understand.

However, I agree that there are rooms for improvement in the current Makefile. I'd be happy to address those in a separate task than to keep this one blocked. I would like to hear both of your thoughts on this. 

cc: @welpaolo

File: .github/workflows/trivy.yml [hunk 5]
```diff
@@ -14,14 +14,14 @@ jobs:
     steps:
       - name: Checkout code
         uses: actions/checkout@v4
-      - name: Install skopeo
+      - name: Install rockcraft (for skopeo)
         run: |
-          sudo snap install --devmode --channel edge skopeo
+          sudo snap install rockcraft --classic --edge
```
Comment (welpaolo) [inline .github/workflows/trivy.yml:19]: Do we need the edge version of rockcraft ?

**Comment by welpaolo** [comment]
I totally agree to have a more simple and readable Makefile for the moment being. We can explore more complex automation in the future.

File: tests/integration/integration-tests-kyuubi.sh [hunk 6]
```diff
@@ -0,0 +1,186 @@
+#!/bin/bash
+
+# The integration tests are designed to tests that Spark Jobs can be submitted and/or shell processes are
```
Comment (welpaolo) [inline tests/integration/integration-tests-kyuubi.sh]: I would update the comment to reflect the Kyuubi tests.

**Review by welpaolo — DISMISSED** [review]
Really good job Bikalpa. 

I would like to have a better separation of base rock image, docker files. I will propose to have subfolder to divide images based on flavors: 
- `charmed-spark`
- `charmed-spark-jupyter`
- `charmed-kyuubi`

I left some minor comments in the PR.

**Review by deusebio — DISMISSED** [review]
I'm happy to proceed with this! Great job! Thanks!

File: .github/workflows/trivy.yml (hunk 5 above)
Comment (theoctober19th) [reply .github/workflows/trivy.yml:19]: Yes, `rockcraft.skopeo` is currently only available on `edge`.

**Review by welpaolo — APPROVED** [review]
LGTM

**Review by deusebio — APPROVED** [review]
LGTM

**Pull request merged**